docs/standard/native-interop/com-wrappers-source-generation.md

```diff
@@ -0,0 +1,166 @@
+---
+title: COM wrappers source generation
+description: Learn about compile-time source generation for COM wrappers in .NET.
+ms.date: 08/25/2023
+---
+
+# Source Generation for COM Wrappers
+
+.Net 8 introduces a source generator that creates an implementation of the [COM wrappers](./com-wrappers.md) API for you. The generator recognizes the <xref:System.Runtime.InteropServices.GeneratedComInterfaceAttribute>.
+
+The .Net runtime's built-in (not source generated) COM interop system generates an IL stub&mdash;a stream of IL instructions that is JIT-ed&mdash;at run time to facilitate the transition from managed code to COM, and vice-versa, on Windows only. Since this IL stub is generated at run time, it isn't compatible with [NativeAOT](../../core/deploying/native-aot/index.md), or [IL trimming](../../core/deploying/trimming/trim-self-contained.md). Generating the IL stub also takes CPU time, and can be difficult to debug when issues arise.
+
+Code using the built in system might look like the following:
+
+```csharp
+[ComImport]
+[Guid("3faca0d2-e7f1-4e9c-82a6-404fd6e0aab8")]
+interface IFoo
+{
+    void Method(int i);
+}
+
+[DllImport("MyComObjectProvider.dll")]
+static nint GetPointerToComInterface();
+
+[DllImport("MyComObjectProvider.dll")]
+static void GivePointerToComInterface(nint comObject);
+
+// Use the system to create a Runtime Callable Wrapper to use in managed code
+nint ptr = GetPointerToComInterface();
+IFoo foo = Marshal.GetObjectForIUnknown(ptr);
+foo.Method(0);
+...
+// Use the system to create a COM Callable Wrapper to pass to unmanaged code
+IFoo foo = GetManagedIFoo();
+nint ptr = Marshal.GetIUnknownForObject(foo);
+GivePointerToComInterface(nint ptr);
+```
+
+The COM wrappers API enables interacting with COM in C# without using the the built-in COM system, but requires lots of boilerplate and hand written unsafe code. The COM interface generator automates that process and makes COM wrappers as easy as built-in COM.
+
+## Basic Usage
```

Comment:
We should have a basic example in dotnet/samples.

Reply:
If you want to start something I can create a consumer side for a full E2E using DNNE and other features.